services/app-api/src/postgres/contractAndRates/replaceRateOnContract.ts

```diff
@@ -0,0 +1,209 @@
+import { type PrismaClient } from '@prisma/client'
+import {
+    type UpdateDraftContractRatesArgsType,
+    updateDraftContractRatesInsideTransaction,
+} from './updateDraftContractRates'
+import { type NotFoundError } from '../postgresErrors'
+import {
+    type UnlockContractArgsType,
+    unlockContractInsideTransaction,
+} from './unlockContract'
+import { type ContractType } from '../../domain-models'
+import {
+    type WithdrawDateArgsType,
+    withdrawRateInsideTransaction,
+} from './withdrawRate'
+import {
+    type SubmitContractArgsType,
+    submitContractInsideTransaction,
+} from './submitContract'
+import { type PrismaTransactionType } from '../prismaTypes'
+
+/**
+ * replaceRateOnContract
+ * @returns contract with updated rates
+ *
+ * unlock, edit, and resubmit contract to unlink a withdrawn rate and use replacement rate instead
+ * finally updates withdrawn rate status
+ * the same update reason is used in all logging (both on the contract unlock/submit info and the rate unlock/submit/withdrawn info)
+ */
+
+type ReplaceRateOnContractArgsType = {
+    contractID: string
+    withdrawnRateID: string
+    replacementRateID: string
+    replacedByUserID: string
+    replaceReason: string
+}
+
+async function replaceRateOnContractInsideTransaction(
+    tx: PrismaTransactionType,
+    args: ReplaceRateOnContractArgsType
+): Promise<ContractType | Error> {
+    const {
+        contractID,
+        replacedByUserID,
+        replaceReason,
+        withdrawnRateID,
+        replacementRateID,
+    } = args
+
+    // unlock contract and rates
+    const unlockContractArgs: UnlockContractArgsType = {
+        contractID,
+        unlockedByUserID: replacedByUserID,
+        unlockReason: replaceReason,
```

Comment:
Should we prefix this reason in some way? I'm imagining someone withdrawing a rate and that reason not quite making sense for a reason for an unlock/submit? like `Withdrawing rate: ${withdrawReason}`

Reply:
Good point. This feels out of scope for this PR and more of a design/product question when we ship the UI.  We can raise this thne. My understanding was they are going to have a very clear reason they will type in